### PR TITLE
boards/arm64/imx9/imx93-evk: add trdc header

### PR DESCRIPTION
**Pull request by joukkone**
Example trdc header for bootloader build

This is needed when building bootloader for imx9-evk

*Note: Please adhere to [Contributing Guidelines](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md).*

## Summary

*Update this section with information on why change is necessary,
 what it exactly does and how, if new feature shows up, provide
 references (dependencies, similar problems and solutions), etc.*

## Impact

*Update this section, where applicable, on how change affects users,
 build process, hardware, documentation, security, compatibility, etc.*

## Testing

*Update this section with details on how did you verify the change,
 what Host was used for build (OS, CPU, compiler, ..), what Target was
 used for verification (arch, board:config, ..), etc. Providing build
 and runtime logs from before and after change is highly appreciated.*

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR does not meet the NuttX requirements. While it provides the template, it lacks the crucial information required by the template.  Specifically, the sections marked with "*Update this section...*" are placeholders and haven't been filled with the actual details of the change.  A PR with these placeholders is incomplete and unreviewable.

Here's a breakdown of what's missing:

* **Summary:**  The summary needs to explain the *why*, *what*, and *how* of the change.  It mentions a trdc header for the imx9-evk bootloader, but doesn't explain *why* this header is needed, *what* it contains, or *how* it solves the problem.  There's no mention of related issues.

* **Impact:**  All the impact fields are left blank.  The author needs to assess and document the impact of the change on users, build process, hardware, documentation, security, and compatibility.

* **Testing:** The testing section is also incomplete.  It requests details about the build host and target, as well as logs from before and after the change, but none of this information is provided.  Simply stating "This is needed" isn't sufficient verification.

To meet the NuttX requirements, the author needs to fill out all the required sections with specific information relevant to the changes made.

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
Hi @joukkone nice contribution! Please include later a board config and a Documentation/ explaining how to use the bootloader on this board

File: boards/arm64/imx9/imx93-evk/include/imx9_trdc_config.h [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+#define X               1
+#define TEE_SHM_SIZE    0x0
+
+/* GLBAC7 is used for TRDC only, any setting to GLBAC7 will be ignored
+ * GLBAC6 is used for fused modules, any setting to GLBAC6 will be ignored
+ */
+
+/* aonmix */
+
+struct trdc_glbac_config trdc_a_mbc_glbac[] =
```
Comment (xiaoxiang781216) [inline boards/arm64/imx9/imx93-evk/include/imx9_trdc_config.h:40]: where we use these new global variables?

File: boards/arm64/imx9/imx93-evk/include/imx9_trdc_config.h [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+{
+  { 0, 0, SP(RW)  | SU(RW)   | NP(RW)  | NU(RW) }, /* MBC0 */
+  { 1, 0, SP(RW)  | SU(RW)   | NP(RW)  | NU(RW) }, /* MBC1 */
+  { 1, 1, SP(RW)  | SU(R)    | NP(RW)  | NU(R)  },
+  { 1, 2, SP(RWX) | SU(RWX)  | NP(RWX) | NU(RWX) },
+};
+
+struct trdc_mbc_config trdc_a_mbc[] =
```
Comment (xiaoxiang781216) [inline boards/arm64/imx9/imx93-evk/include/imx9_trdc_config.h:48]: add g_ prefix, should we move all global variables to c file

**Comment by joukkone** [comment]
Updated patch. There is now full trdc configuration, Those header files can be generated with that nxp chiptool, Trdc configuration must be done before DDR configuration. Final configuration should be executed just before jumping to EL1.

File: boards/arm64/imx9/imx93-evk/include/imx9_trdc_config.h (hunk 2 above)
Comment (joukkone) [reply boards/arm64/imx9/imx93-evk/include/imx9_trdc_config.h]: I changed those static, these are board specific and only used by imx9_trdc.c

File: arch/arm64/src/imx9/hardware/imx9_trdc.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+#define ELE_MAX_MSG 255U
+#define AHAB_VERSION 0x6
+#define AHAB_CMD_TAG 0x17
+#define AHAB_RESP_TAG 0xe1
+#define ELE_RELEASE_RDC_REQ (0xC4)
+#define ELE_READ_FUSE_REQ	  (0x97)
+#define ELE_OK            0xd6
+
+#define FSB_BASE			0x47510000UL
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/hardware/imx9_trdc.h:49]: align the second column to the same position

File: arch/arm64/src/imx9/imx9_trdc.c [hunk 4]
```diff
@@ -0,0 +1,929 @@
...
+#include "chip.h"
+#include "arm64_internal.h"
+#include "imx9_trdc.h"
+#include <arch/board/imx9_trdc_config.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define ARRAY_SIZE(a)				\
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/imx9_trdc.c:45]: remove use nitems

File: arch/arm64/src/imx9/imx9_trdc.c [hunk 5]
```diff
@@ -0,0 +1,929 @@
...
+  (sizeof(a) / sizeof((a)[0]))
+
+#define mmio_read_32(c)	getreg32(c)
+#define mmio_write_32(c, v)	putreg32(v, c)
+
+#define mmio_clrbits_32(addr, clear) modifyreg32(addr, clear, 0)
+#define mmio_setbits_32(addr, set) modifyreg32(addr, 0, set)
+#define mmio_clrsetbits_32(addr, clear, set) modifyreg32(addr, clear, set)
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/imx9_trdc.c:53]: align the second column to the same position

File: arch/arm64/src/imx9/imx9_trdc.c [hunk 6]
```diff
@@ -0,0 +1,929 @@
...
+    }
+
+  return -EIO;
+}
+
+/****************************************************************************
+ * Trdc_from atf
+ ****************************************************************************/
+
+struct mbc_mem_dom
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/imx9_trdc.c:211]: move all type to private type section

File: arch/arm64/src/imx9/imx9_trdc.c [hunk 7]
```diff
@@ -0,0 +1,929 @@
...
+  uint8_t blk_num;
+};
+
+struct trdc_fuse_data
+{
+  uint8_t fsb_index;
+  uint32_t value;
+};
+
+static const struct trdc_mgr_info trdc_mgr_blks[] =
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/imx9_trdc.c:314]: move all global to private data section

File: arch/arm64/src/imx9/imx9_trdc.c [hunk 8]
```diff
@@ -0,0 +1,929 @@
...
+{
+  { 0x44270000, 0, 0, 39, 40 }, /* TRDC_A */
+  { 0x42460000, 0, 0, 70, 71 }, /* TRDC_W */
+  { 0x42460000, 1, 0, 1,  2  }, /* TRDC_M */
+  { 0x49010000, 0, 1, 1,  2  }, /* TRDC_N */
+};
+
+static const struct trdc_config_info trdc_cfg_info[] =
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/imx9_trdc.c:322]: add g_ prefix for ALL global variables

File: arch/arm64/src/imx9/hardware/imx9_trdc.h (hunk 3 above)
Comment (joukkone) [reply arch/arm64/src/imx9/hardware/imx9_trdc.h]: done

File: arch/arm64/src/imx9/imx9_trdc.c (hunk 4 above)
Comment (joukkone) [reply arch/arm64/src/imx9/imx9_trdc.c]: done

File: arch/arm64/src/imx9/imx9_trdc.c [hunk 9]
```diff
@@ -0,0 +1,938 @@
...
+  {
+    0x49010000,
+    trdc_boot_mbc_glbac, nitems(trdc_boot_mbc_glbac),
+    trdc_boot_ocram_mbc, nitems(trdc_boot_ocram_mbc),
+    trdc_n_mrc_glbac, nitems(trdc_n_mrc_glbac),
+    trdc_n_mrc, nitems(trdc_n_mrc)
+  },
+};
+
+/****************************************************************************
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/imx9_trdc.c:215]: remove dup banner

File: arch/arm64/src/imx9/imx9_trdc.c [hunk 10]
```diff
@@ -0,0 +1,938 @@
...
+
+#define mmio_read_32(c)	                      getreg32(c)
+#define mmio_write_32(c, v)                   putreg32(v, c)
+#define mmio_clrbits_32(addr, clear)          modifyreg32(addr, clear, 0)
+#define mmio_setbits_32(addr, set)            modifyreg32(addr, 0, set)
+#define mmio_clrsetbits_32(addr, clear, set)  modifyreg32(addr, clear, set)
+
+/* Bits 8:15 rdc id, bits 0:7 core id */
+
+#define TRDC_AON            0x7402
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/imx9_trdc.c:53]: align ALL second column

File: arch/arm64/src/imx9/hardware/imx9_trdc.h [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+#define ELE_MU_TSR	(IMX9_S3MUA_BASE+ 0x124)
+#define ELE_MU_RCR	(IMX9_S3MUA_BASE+ 0x128)
+#define ELE_MU_RSR	(IMX9_S3MUA_BASE+ 0x12c)
+
+#define ELE_RR_NUM        4
+#define ELE_TR_NUM        8
+#define ELE_MU_TR(i)	(IMX9_S3MUA_BASE + 0x200 + (i) * 4)
+#define ELE_MU_RR(i)	(IMX9_S3MUA_BASE + 0x280 + (i) * 4)
+
+#define DID_NUM 16
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/hardware/imx9_trdc.h:64]: align ALL second column to the same position

File: arch/arm64/src/imx9/hardware/imx9_trdc.h [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+#define IMX9_MBC_MEM_BLK_CFG_0(m, n)    (0x200 * m + 0x40 + (n << 2))
+#define IMX9_MBC_MEM_BLK_CFG_I(m, n, i) (0x200 * m + 0x40 + (80 << 2) + (i - 1) * 0x28 + (n << 2))
+#define IMX9_MRC0_DOM_RGD_W(m, n)       (0x100 * m + 0x40 + (n << 3))
+
+#define ELE_MAX_MSG             255U
+#define AHAB_VERSION            0x6
+#define AHAB_CMD_TAG            0x17
+#define AHAB_RESP_TAG           0xe1
+#define ELE_RELEASE_RDC_REQ     0xc4
+#define ELE_READ_FUSE_REQ	      0x97
```
Comment (xiaoxiang781216) [inline arch/arm64/src/imx9/hardware/imx9_trdc.h:46]: ditto

**Comment by xiaoxiang781216** [comment]
@joukkone let's squash the patch into one.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**